extensions/mapstore/src/main/java/com/hazelcast/mapstore/GenericMapLoader.java

```diff
@@ -0,0 +1,164 @@
+import com.hazelcast.sql.SqlService;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+
+class GenericMapLoader<K> implements MapLoader<K, GenericRecord> {
+
+    protected SqlService sql;
+
+    protected Queries queries;
+
+    protected GenericMapStoreProperties properties;
+
+    @Nonnull
+    private GenericRecord convertRowToGenericRecord(SqlRow row) {
+        GenericRecordBuilder builder = GenericRecordBuilder.compact(properties.compactTypeName);
+
+        SqlRowMetadata metadata = row.getMetadata();
+        for (int i = 0; i < metadata.getColumnCount(); i++) {
+            SqlColumnMetadata column = metadata.getColumn(i);
+
+            if (column.getName().equals(properties.idColumn) && !properties.idColumnInColumns) {
+                continue;
+            }
+
+            switch (column.getType()) {
+                case VARCHAR:
+                    builder.setString(column.getName(), row.getObject(i));
+                    break;
+
+                case BOOLEAN:
+                    builder.setBoolean(column.getName(), row.getObject(i));
+                    break;
+
+                case TINYINT:
+                case SMALLINT:
+                case INTEGER:
+                    builder.setInt32(column.getName(), row.getObject(i));
+                    break;
+
+                case BIGINT:
+                    builder.setInt64(column.getName(), row.getObject(i));
+                    break;
+
+                case DECIMAL:
+                    builder.setDecimal(column.getName(), row.getObject(i));
+                    break;
+
+                case REAL:
+                    builder.setFloat32(column.getName(), row.getObject(i));
+                    break;
+
+                case DOUBLE:
+                    builder.setFloat64(column.getName(), row.getObject(i));
+                    break;
+
+                case DATE:
+                    builder.setDate(column.getName(), row.getObject(i));
+                    break;
+
+                case TIME:
+                    builder.setTime(column.getName(), row.getObject(i));
+                    break;
+
+                case TIMESTAMP:
+                    builder.setTimestamp(column.getName(), row.getObject(i));
+                    break;
+
+                case TIMESTAMP_WITH_TIME_ZONE:
+                    builder.setTimestampWithTimezone(column.getName(), row.getObject(i));
+                    break;
+
+                default:
+                    throw new HazelcastException("Column type " + column.getType() + " not supported");
+            }
+        }
+
+        return builder.build();
+    }
+
+    protected void initMapLoader(HazelcastInstance instance, Properties properties, String mapName) {
+        this.properties = new GenericMapStoreProperties(properties, mapName);
+        sql = instance.getSql();
+    }
```

Comment:
The `GenericMapLoader` should also implement the `MapLoaderLifecycleSupport`, this should go into the init method.

Reply:
GenericMapLoader  now inherits from MapLoaderLifecycleSupport